chore: add everything (#1)
thepassle/batching-blog: README.md

> # Batched updates blog

Reviewer: Reactive Updates without VDOM

Asynchronous Rendering for Web Components

Efficient DOM Updates without VDOM

Efficient DOM Updates for Web Components

Reviewer 2: Building a reactive front end framework in 50 LOC ?
Barebones reactive web components
A primer on Web Components rerendering

Reviewer 2: Not Every Hero Wears A Mask

![image](https://user-images.githubusercontent.com/921666/100021151-56146200-2de1-11eb-899d-bcfca6f7dee9.png)

Gotta come up with a clever title. Suggestions?

> Whats always better than doing work? Not doing work!

## Intro

> [LitElement](https://lit-element.polymer-project.org/) is a base class for developing web components. It's very lightweight, very fast at what it does, and takes a lot of the heavy lifting of writing web components out of the developers hands, by being lazy (or: efficient).

Reviewer 3: If there were likes to reference `lightweight` and `very fast` you could be the change in the JS community that stops blankly stating these sorts of things without context. _No pressure 😏 _

Some questions that I often get during workshops, or from new colleagues who have just started working with LitElement are: 
- How exactly does LitElement achieve its efficient updates? 
- What does asynchronous rendering mean? 
- How does LitElement pick up property changes?

So I figured, _it's blog time!_

Lets start by taking a look at the following code example:

```js
class MyElement extends LitElement {
  foo() {
    this.myPropertyA = 1;
  }
}
```

> LitElement is _reactive_ and _observes_ your properties to kick off what is known as its "render pipeline". What this means is that whenever we change a property in our component, it'll update and re-render the component. This is nice, because it means we don't have to do anything manually. In the code example above, just the act of setting a new value to `this.myPropertyA` will cause our component to re-render.

Reviewer 3: "Reactive" and "observes" are either the _same_ or technically different...right?

Also, it's too bad you missed the `ReactiveElement` conversation last week with the Polymer team.

Reviewer 3: Also, `setting a new value` maybe you get into this later, but is it worth talking about identity checking, etc?

Fortunately, LitElement is smart and lazy, and uses a clever technique to _batch_ these updates, and only re-render _once_.

## Batching

Before we get started, lets make clear what we actually mean when we talk about _batching work_, and go over a couple techniques to demonstrate how we can achieve batching behavior in JavaScript.

Batching is basically limiting the amount of work we need to do, and is often implemented as a performance measure. A popular usecase for batching, or limiting work, is limiting API calls that may get fired in quick succession for example. We don't want to put any unnecessary load on the API, and do a bunch of wasteful requests. Or in the case of LitElement; efficiently limiting re-renders of our components. Remember; the goal is to avoid _any_ unnecessary work in order to be performant.

> A popular pattern for batching is using a debounce function. As mentioned above, debouncers are often used for batching API calls. Imagine we have a search input, that every time a new search keyword is entered, fetches some search results from a server. It would put a pretty heavy load on the API if we would fire a request for every single input change or every new character, and that can be pretty wasteful. Instead, we can _debounce_ the API calls, to only actually fire a request on the _last_ input change.

Reviewer: maybe link to some debounce posts. scotch.io has a nice one iirc

Consider the following example:

> ```js

Reviewer: Maybe instead of a code snippet, use a diagram to explain debounce, since it's not the main focus of the article, and code snippets (even stunningly beautiful ones like this) tend to tire the reader

});

console.log(3);
```

Here's the answer, the order of the resulting console logs of this code is:

```js
// 1
// 3
// 2
```

We'll have [Jake Archibald](https://twitter.com/jaffathecake) explain to us why exactly this happens:

> The microtask queue is processed after callbacks as long as no other JavaScript is mid-execution, and at the end of each task. Any additional microtasks queued during microtasks are added to the end of the queue and also processed.

If we go back to our code snippet, we can see that the following happens:

- `console.log(1);` is called, and logged to the console
- We schedule a new _microtask_ with `Promise.resolve().then(() => {})`, but at this point in time, JavaScript has not finished executing; it's not time to actually process microtasks just yet
- JavaScript continues executing code, because we're not done yet
- `console.log(3);` is called, and logged to the console
- JavaScript has finished executing! The browser can now process any microtasks we may have
- `console.log(2);` is called, and logged to the console.

So why is this important? How does this allow us to use microtasks to _batch_ work? Let's take a look at a practical example:

```js

> /* We declare and initialize a variable to keep track of whether or not an update has already been requested */

Reviewer: since you're going to later on anyways, why not adopt class syntax already here?

let updateRequested = false;

function scheduleUpdate() {
  /** 
   * In here, we need a check to see if an update is already previously requested.
   * If an update already is requested, we don't want to do any unnecessary work!
   */
  if(!updateRequested) {

    /* If no update has yet been requested, we set the `updateRequested` flag to `true` */
    updateRequested = true;

    /** 
     * Since we now know that microtasks run after JavaScript has finished
     * executing, we can use this knowledge to our advantage, and only set 

>      * the `updateRequested` flag back to `false` again once all the tasks 

Reviewer: this comment is the heart of the matter. consider putting it in the post body below, not the snippet

     * have run, essentially delaying, or _batching_ the update!
     */
    Promise.resolve().then(() => {
      updateRequested = false;
      update();
    });
  }
}

/* This is our `update` function that we only want to be called once */
function update() {
  console.log('updating!');
}

/* We call scheduleUpdate in quick succession */
scheduleUpdate();
scheduleUpdate();
scheduleUpdate();

// "updating!" only logged once
```

> What we do here is essentially _gatekeep_ potentially multiple incoming calls for `scheduleUpdate` with the `if` statement, and cleverly use our newfound knowledge of microtasks to our advantage to process any other incoming calls _first_, and only once JavaScript has finished executing, and when its microtask processing time, our `update` function is called.

Reviewer: ```suggestion
Essentially, `scheduleUpdate` guards the `update` function against multiple incoming calls with its `if` statement. This cleverly uses our newfound knowledge of microtasks to our advantage to process any other incoming calls _first_. `update` only gets called once JavaScript has finished executing and microtask processing begins.
```

As a fun tidbit of knowledge, we can also write the `scheduleUpdate` function like so:

```js

> function scheduleUpdate() {

Reviewer: ```suggestion
async function scheduleUpdate() {
```

  if(!updateRequested) {
    updateRequested = true;
    updateRequested = await false;
    update();
  }
}
```

Since, from [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await):

> > If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise.

Reviewer: ```suggestion
> If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise.

In otherwords, once we set `updateRequested`, we don't immediately unset, the `await` keyword ensures that we only _schedule_ it to be unset. 
```

Reviewer: Something i've been wondering is whether  each `await` keyword schedules a new microtask, and what the ramifications of the answer are. maybe worth a footnote

In fact, [Justin Fagnani](https://twitter.com/justinfagnani), one of the authors of LitElement, once told me on twitter that the pre-LitElement version did exactly this:

![twitter screenshot](./img/justin.png)

### Deferring promises

> Alright, we're almost getting to the good stuff now. But before we go there, there's one final concept we need to explore, and thats deferring resolving of Promises, which will provide us with a handy utility later on in this blog post.

Reviewer: ```suggestion
Alright, we're almost ready to explain how LitElement achieves its efficient updates. But before we go there, there's one final concept we need to explore, and that's deferring resolving of Promises, which will provide us with a handy utility later on in this blog post.
```

I already think this is good stuff :D

Take a look at the following code:

```js
/* We declare a new variable */
let resolveFn;

/** 
 * We declare a new promise, but we _dont_ resolve it! 
 * Instead, we assign the `resolve` function to the
 * `resolveFn` that we declared above
 */
const myPromise = new Promise((resolve) => {
  resolveFn = resolve;
});

/** 
 * `myPromise` now is just a pending promise, waiting to be resolved,
 * and we can do that whenever we please with our `resolveFn`
 */
myPromise; // Promise {<pending>}

/* For example, we could do a whole bunch of other work in between */

/* And finally, whenever we're ready, we can call the `resolveFn` to resolve `myPromise` */
resolveFn(); // Resolves `myPromise`

/* Aaand `myPromise` is now fulfilled! */
myPromise; // Promise {<fulfilled>: undefined} ! 🎉
```

This pattern is often used in cases (like unit tests) where you have to await a specific amount of time before continuing.

```js

> function sleep(delay) {

Reviewer: I'm not sure sleep is the best example, because you're passing resolve directly in the body of the promise, where as with `updateComplete` the fun starts when you reassign resolve to something out of scope

  return new Promise(resolve => {
    setTimeout(resolve, delay)
  });
}

(async () => {
  await sleep(1000); // Awaits one second for the timeout to execute before continuing
})()
```

In this example, the `sleep` function returns a new `Promise`. Inside the promise, we set a timeout, but we only resolve the promise once the `setTimeout` callback is executed. So by `await`ing the `sleep` function, we essentially force to wait until `setTimeout` callback has executed.

Be aware, however, that JavaScript is single-threaded, and blocking the main thread is generally considered a bad practice.

## Back to LitElement

### 🧙‍♂️ Automagically requested updates

LitElement is _reactive_. What this means is that, like many other modern frontend libraries, we dont have to worry as much about manually rendering or re-rendering the contents of our component. All we have to do as a user of LitElement is declare our properties, LitElement will then _observe_ these properties for us, and when a property has changed LitElement takes care of updating our component for us. Nice. I like not doing work.

> ![counter component](./img/wcdev.png)

Reviewer: <plug>
If you do this, than any page with the element registered (like ours, when we run the build we'll detect that the element is used and append) will get a live preview, but everyone else will get the image. 
</plug>

```suggestion
<wcd-live title="Counter Component">

[![counter component](./img/wcdev.png)](https://webcomponents.dev/edit/THE_ID)

</wcd-live>
```

Author: How would this work/behave on medium? 🤔

Reviewer: since medium doesn't support markdown, whatever tool you're using would just strip out the tags

https://markdowntomedium.com/

> Simple counter element, courtesy of the lovely folks over at [webcomponents.dev](https://webcomponents.dev)

> Imagine we have a simple counter element that has a `count` property. All we need to do to trigger LitElement to rerender is set a new value to the `count` property: `this.count = 1;`, and LitElement will automagically _request_ an update.

Reviewer 3: This sentence is a little hard to consume: "All we need to do to trigger LitElement to rerender is set a new value to the `count` property:"

But... how? How does LitElement know that the `count` property was set? How does it _observe_ our properties?

> LitElement requires users to declare the properties of their component in a _static properties getter_. This static properties getter is very, very useful, because it takes a lot of boilerplate out of the developers hands, it takes care of attributes for us, handles attribute reflection, allows us to react to changes, but more importantly: It creates getters and setters for us. This is nice, because this means that our components dont get littered with getters and setters all over the place, but it also requests updates for us!

Reviewer 3: Might be worth linking to one of you or Benny's great articles on the difference between vanilla attribute observing and what LitElement does here...

Consider the following example:
```js
class MyCounter extends LitElement {
  static get properties() {
    return {
      count: { type: Number }
    }
  }
}
```

Under the hood, LitElement essentialy turns this into the following getters/setters:

```js
class MyCounter extends LitElement {
  set count(value) {
    this.__count = value;

>     this.requestUpdate(); // ❗️ (metal-gear-solid-alert.mp3)

Reviewer: 😄

Reviewer 2: hahaha

  }

  get count() {
    return this.__count;
  }
}
```

> Note: this is pseudocode for illustration purposes, the actual source code of LitElement looks a little bit different, but the general concept still applies. If you're interested in reading the source code, you can find it [here](https://github.com/Polymer/lit-element/blob/master/src/lib/updating-element.ts#L361).

What this means is that whenever we assign a new value to a property like so: `this.count = 1;`, the setter for `count` is called, which then calls `requestUpdate()`. So by simply setting a property, an update is requested for us!

### Objects and arrays

> We'll take a small aside here to explain another common occurence when using LitElement, that I often see people get confused about. We now know that we can trigger updates simply by setting a new value to a property, but... what about objects and arrays?

Reviewer 3: 🔥

Imagine we have a component with a `user` property that looks something like this:

```js
{
  name: "Mark",
  age: 30
}
```

> People often expect that setting, for example, the `name` property on the `user` object should trigger a rerender, like so: `this.user.name = "Nick";`, and are surprised to find that it _doesn't_. The reason for this is that by setting the `name` property of the `user` object, we don't actually change the value of the `user` object itself, and as such, the `requestUpdate()` method is never called!

Reviewer 3: Can we really say the `value` doesn't change here? I always use the word `identity` here, but if we can correctly use `value` instead, that would be much easier for me to explain to people.

> People often expect that setting, for example, the `name` property on the `user` object should trigger a rerender, like so: `this.user.name = "Nick";`, and are surprised to find that it _doesn't_. The reason for this is that by setting the `name` property of the `user` object, we don't actually change the value of the `user` object itself, and as such, the `requestUpdate()` method is never called!

Reviewer: ```suggestion
People often expect that setting, for example, the `name` property on the `user` object should trigger a rerender, like so: `this.user.name = "Nick";`, and are surprised to find that it _doesn't_. The reason for this is that by setting the `name` property of the `user` object, we don't actually change the reference to the `user` object itself, only the value of one of it's properties. As such, the `requestUpdate()` method is never called!
```

> An easy fix for this is to just call `this.requestUpdate()` manually, or alternatively replace the entire object to make sure the setter _does_ get called:

Reviewer 3: Sometimes you use `this.requestUpdate()` and other `requestUpdate()`, not sure if you want to normalize that.

To do so, we're going to write a very naive implementation of this behavior. We're going to start with a very simple JavaScript class that efficiently calls an `update` method when our properties change. Follow along:

```js
/* We create a new class ... */
class Batching {

  /* ... that has a `requestUpdate` method */
  async requestUpdate() {

    /** 
     * In here, we need a check to see if an update has already previously been requested.
     * If an update is already requested, we don't want to do any unnecessary work!
     */
    if (!this.updateRequested) {

      /**
       * If no update was previously requested, we set this
       * flag to `true` to avoid doing unnecessary work in
       * case `requestUpdate` might get called another (or several) times
       */
      this.updateRequested = true;

      /** 
       * ... and _this_ is where the magic happens;
       * This schedules a microtask that executes once JavaScript has
       * finished executing, and since we guard against any other
       * incoming calls for `requestUpdate`, we only do work once

>        */

Reviewer: maybe leave comments here like `STEP 1` etc and do the text in the post

  }
}

/* We instantiate a new instance of our class */
const batching = new Batching();

/* And we set multiple properties */
batching.a = 1;
batching.b = 2;

/** 
 * 🎉 The result:
 * "updating!" is only logged to the console once because of the batching!
 */
```

### Taking things a step further

Neat! We now have a simple implementation of batching updates. We can now take things even further and provide some kind of API so users can `await` updates, like LitElement's `updateComplete` property. Take a look at the following code:

```js
class Batching {
  constructor() {

    /** 
     * We initialize an `updateComplete` property with a new 
     * Promise that we'll only resolve once `this.update()`
     * has been called
     */
    this.updateComplete = this.__createDeferredPromise();
  }

  async requestUpdate() {
    if (!this.updateRequested) {
      this.updateRequested = true;
      this.updateRequested = await false;
      this.update();

      /* When our update is, in fact, complete we resolve the Promise that was assigned to the `updateComplete` property ... */
      this.__resolve();

      /* ... And we assign a new promise to updateComplete for the next update */

>       this.updateComplete = this.__createDeferredPromise();

Reviewer 3: Does lit element really pre-set an unresolved promise? I guess maybe because it `update()`s in response to `construction/attachement`? Interesting... 🤔

Reviewer: worth the price of admission, right?

Author: yep https://github.com/Polymer/lit-element/blob/master/src/lib/updating-element.ts#L525

    }
  }

  update() {
    console.log("updating!");
  }

  /** 
   * Creates a new deferred promise that we can await, 
   * and assign the `resolve` function to `this.__resolve`, 
   * so we can resolve the promise after we call `this.update()`
   */
  __createDeferredPromise() {
    return new Promise((resolve) => {
      this.__resolve = resolve;
    });
  }

  set a(val) {
    this.__a = val;
    this.requestUpdate();
  }

  set b(val) {
    this.__b = val;
    this.requestUpdate();
  }
}

/* We use an IIFE (Immediately Invoked Function Expression), because top-level await is not a thing yet 😑 */
(async () => {
  /* We instantiate a new instance of our class */
  const batching = new Batching();

  /* Set multiple properties in a row */
  batching.a = 1;
  batching.b = 2;

  /* And this is where we `await` an update */
  await batching.updateComplete;

  /* We then assign another property */
  batching.b = 3;

  /** 
   * 🎉 The result:

>    * "updating!" is logged to the console twice!

Reviewer 3: You might want to structure this and the other version(s) as an actual console so that they see all of the output and don't need to process `is logged to the console twice!` to fully grasp what happened.

    });
  }
}
```

> Again, this code is a simplified example of the technique LitElement uses for illustration purposes, if you're interested in seeing how LitElement achieves this exactly, you can find the source code [here](https://github.com/Polymer/lit-element/blob/master/src/lib/updating-element.ts#L715).

And, well, that's it. Not much has changed here really, other than the fact that we now extend from `HTMLElement`, and have to call `super()` in the constructor. Let's see how we can actually use it in a component:

```js
/* We create a new class, MyElement, that extends from our BatchingElement base class */
class MyElement extends BatchingElement {

  constructor() {
    super();

    /* We attach a shadowRoot to our component for good measure */
    this.attachShadow({ mode: "open" });
  }

  connectedCallback() {
    /* And we call an initial render when our component gets connected to the DOM */
    this.update();
  }

  /**
   * This is our `update` function that will get triggered by the
   * `requestUpdate` method.
   * 
   * Any time we set a property on this element, we'll also
   * trigger an update.
   */
  update() {
    console.log('updating!');

>     this.shadowRoot.innerHTML = `

Reviewer: yeah seeing this again gives me twitter flashbacks. I say never put `shadowRoot.innerHTML = ` in a public post even if it' explicitly a naive implemention

# Batched updates blog

> Gotta come up with a clever title. Suggestions?

Reviewer 2: BTW it might be tainted with context, but is batching the word being used a lot for that use case? To me batching is much more API related. Curious

Fortunately, LitElement is smart and lazy, and uses a clever technique to _batch_ these updates, and only re-render _once_.

## Batching

Before we get started, lets make clear what we actually mean when we talk about _batching work_, and go over a couple techniques to demonstrate how we can achieve batching behavior in JavaScript.

Batching is basically limiting the amount of work we need to do, and is often implemented as a performance measure. A popular usecase for batching, or limiting work, is limiting API calls that may get fired in quick succession for example. We don't want to put any unnecessary load on the API, and do a bunch of wasteful requests. Or in the case of LitElement; efficiently limiting re-renders of our components. Remember; the goal is to avoid _any_ unnecessary work in order to be performant.

> A popular pattern for batching is using a debounce function. As mentioned above, debouncers are often used for batching API calls. Imagine we have a search input, that every time a new search keyword is entered, fetches some search results from a server. It would put a pretty heavy load on the API if we would fire a request for every single input change or every new character, and that can be pretty wasteful. Instead, we can _debounce_ the API calls, to only actually fire a request on the _last_ input change.

Reviewer 2: reading the article again, this example does not bring anything (technically) for the article. It's a nice example, but with a different implementation ?

Author: Debouncing/throttling a very popular ways of 'batching' or deferring work in front-end-land, when you mention 'batching' or 'scheduling', I think most people will instantly jump to: "Oh, so debouncing.". I think its an important and popular enough pattern to briefly explain and give context, and also to clarify that this is _not_ how LitElement does it

Consider the following example:

```js
/* This is generally what a debounce function looks like, it takes a function to execute, and optionally a delay */

> function debounce(func, delay = 0) {

Reviewer 2: It's minor, but your comments in the code don't read so well in their current state. Not sure where you plan on publishing that : 

<img width="895" alt="image" src="https://user-images.githubusercontent.com/921666/100018987-a12c7600-2ddd-11eb-9526-89bf503b893b.png">

  return function() {

    /* Because this function closes over the outer function, we'll still have access to the `timeoutId` declared above. This is useful on subsequent calls to the 'inner' function. */
    clearTimeout(timeoutId);

    /* We then schedule a new timeout, that calls the function that we'd like to execute */
    timeoutId = setTimeout(() => {
      func.apply(this, arguments);
    }, delay);
  }
};

function search() {
  /* Do some API call to get more search results */
}

/* We create a new variable that stores the _returned_ function from the debounce function */
const debouncedSearch = debounce(search, 500);
input.addEventListener("input", debouncedSearch);
```

If the user types the search keyword "javascript" in the input, instead of firing 10 API requests (one request for every character in the word "javascript"), we only fire one request; when the user has finished typing.

> If you're interested in reading more about debouncing, here's a great blog on [css-tricks.com](https://css-tricks.com/debouncing-throttling-explained-examples/)

> However, and I’m sorry to disappoint, this is not how LitElement does things, but important to illustrate the point of "batching". Before we go deeper into that, we’re going to require some knowledge about...

Reviewer 2: there are many many commas in not so much text in this paragraph :)

## THE EVENT LOOP!

The event loop... can be a pretty elusive subject to understand. I'll try to keep it brief here with a basic example, but if you're interested in learning more about how the event loop works, here are some incredible resources that I cannot recommend enough:

> - [Tasks, Microtasks, Queues and Schedules](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) by [Jake Archibald](https://twitter.com/jaffathecake)

Reviewer 2: This is not specific to this part in particular, but for the whole article. You build up a story. To do this, you lay a lot of groundwork. But those are not directly related to each other.
To me, it was a little confusing because I don't know where you're bringing me. 

I wonder if it would help in the intro to lay all the pieces you'll be going over in the form of a diagram, or a TOC or something.

Reviewer 2: It's even more true as I go down the article.

web components -> batching -> arrays / objects -> batching -> web components.

```js
function sleep(delay) {
  return new Promise(resolve => {
    setTimeout(resolve, delay)
  });
}

(async () => {
  await sleep(1000); // Awaits one second for the timeout to execute before continuing
})()
```

In this example, the `sleep` function returns a new `Promise`. Inside the promise, we set a timeout, but we only resolve the promise once the `setTimeout` callback is executed. So by `await`ing the `sleep` function, we essentially force to wait until `setTimeout` callback has executed.

Be aware, however, that JavaScript is single-threaded, and blocking the main thread is generally considered a bad practice.

## Back to LitElement

### 🧙‍♂️ Automagically requested updates

LitElement is _reactive_. What this means is that, like many other modern frontend libraries, we dont have to worry as much about manually rendering or re-rendering the contents of our component. All we have to do as a user of LitElement is declare our properties, LitElement will then _observe_ these properties for us, and when a property has changed LitElement takes care of updating our component for us. Nice. I like not doing work.

> ![counter component](./img/wcdev.png)

Reviewer 2: I stupidly tried to click that stuff. :D

   * LitElement makes this for us instead? 😩
   */
  set a(val) {
    this.__a = val;
    this.requestUpdate();
  }
  get a(){
    return this.__a;
  }

  set b(val) {
    this.__b = val;
    this.requestUpdate();
  }
  get b(){
    return this.__b;
  }
}

customElements.define("my-element", MyElement);
```

Neat! We've now implemented our `BatchingElement` base class in a real life web component. Whenever we change multiple properties in a row, we'll only call our `update` method once. Consider the following example:

```js
/* We get a reference to our component, so we can call methods on it */
const myElement = document.querySelector('my-element');

/**
 * We assign multiple properties in quick succession,
 * but our component will still only call `update` once!
 */
myElement.a = 'foo';
myElement.b = 'bar';

/* 🎉 The result: */

// "updating!"
```

If you're interested in seeing another real world example that implements this pattern, you can take a look at [`@generic-components`](https://github.com/thepassle/generic-components/blob/master/utils/BatchingElement.js). The implementation is slightly different, but the same concepts apply.

> And thats how LitElement efficiently batches updates and avoids doing unnecessary work. Combine that with [lit-html](https://lit-html.polymer-project.org/) to do efficient DOM updates, and you have a really efficient and powerful combination of libraries at your disposal.

Reviewer 2: It might be worth just mentioned what generic components tries to achieve. It's a bit of a dry link as a conclusion mentioned like that imho

Reviewer 2: maybe also worth coming back to the people you mentioned at the beginning or the article?

Reviewer 2: btw as I was reading I was wondering about actual performance changes. Im guessing that in all the examples you mentioned it wouldn't really matter, but as the dom gets deep it does make a huge difference. might be worth mentioning?

Author: Yeah the 'outro' is a bit abrupt right now, too. I'll try to flesh it out a bit more
